File: ios/Pbkdf2.m
#import "Pbkdf2.h"
#import <CommonCrypto/CommonCrypto.h>

@implementation Pbkdf2

RCT_EXPORT_MODULE()

// Example method
// See // https://reactnative.dev/docs/native-modules-ios
RCT_REMAP_METHOD(derive,
                 withPassword:(nonnull NSString*)password
                 withSalt:(nonnull NSString*)salt
                 withRounds:(nonnull NSNumber*)rounds
                 withKeyLength:(nonnull NSNumber*)keyLength
                 withHash:(nonnull NSString*)hash
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    
    NSMutableData *derivedKey = [NSMutableData dataWithLength: [keyLength unsignedIntValue]];
    NSData *passwordData = [[NSData alloc] initWithBase64EncodedString:password options:0];
    NSData *saltData = [[NSData alloc] initWithBase64EncodedString:salt options:0];
    
    CCPseudoRandomAlgorithm prf = kCCPRFHmacAlgSHA1;
    if ([hash isEqualToString:@"sha-512"]) {
        prf = kCCPRFHmacAlgSHA512;
    }
    if ([hash isEqualToString:@"sha-256"]) {
        prf = kCCPRFHmacAlgSHA256;
    }

    CCKeyDerivationPBKDF(kCCPBKDF2,               // algorithm
                         passwordData.bytes,           // password
                         passwordData.length,          // passwordLength
                         saltData.bytes,              // salt
                         saltData.length,             // saltLen
                         prf,       // PRF
                         [rounds unsignedIntValue],                  // rounds
                         derivedKey.mutableBytes, // derivedKey
                         derivedKey.length);      // derivedKeyLen

    resolve([derivedKey base64EncodedStringWithOptions:0]);
}

@end

File: ios/Pbkdf2.swift
@objc(Pbkdf2)
class Pbkdf2: NSObject {

    @objc(derive:withSalt:withIterations:withHash:withKeyLength:withResolver:withRejecter:)
    func derive(password: String, salt: Data, iterations: Int, hash: String, keyLength: Int, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        if (hash == "sha-1") {
            resolve(self.pbkdf2SHA1(password: password, salt: salt, keyByteCount: keyLength, rounds: iterations))
        } else if (hash == "sha-256") {
            resolve(self.pbkdf2SHA256(password: password, salt: salt, keyByteCount: keyLength, rounds: iterations))
        } else if (hash == "sha-512") {
            resolve(self.pbkdf2SHA512(password: password, salt: salt, keyByteCount: keyLength, rounds: iterations))
        }
    }
    
    func pbkdf2SHA1(password: String, salt: Data, keyByteCount: Int, rounds: Int) -> Data? {
        return pbkdf2(hash:CCPBKDFAlgorithm(kCCPRFHmacAlgSHA1), password:password, salt:salt, keyByteCount:keyByteCount, rounds:rounds)
    }

    func pbkdf2SHA256(password: String, salt: Data, keyByteCount: Int, rounds: Int) -> Data? {
        return pbkdf2(hash:CCPBKDFAlgorithm(kCCPRFHmacAlgSHA256), password:password, salt:salt, keyByteCount:keyByteCount, rounds:rounds)
    }

    func pbkdf2SHA512(password: String, salt: Data, keyByteCount: Int, rounds: Int) -> Data? {
        return pbkdf2(hash:CCPBKDFAlgorithm(kCCPRFHmacAlgSHA512), password:password, salt:salt, keyByteCount:keyByteCount, rounds:rounds)
    }

    func pbkdf2(hash :CCPBKDFAlgorithm, password: String, salt: Data, keyByteCount: Int, rounds: Int) -> Data? {
        let passwordData = password.data(using:String.Encoding.utf8)!
        var derivedKeyData = Data(repeating:0, count:keyByteCount)

        let derivationStatus = derivedKeyData.withUnsafeMutableBytes {derivedKeyBytes in
            salt.withUnsafeBytes { saltBytes in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    password, passwordData.count,
                    saltBytes, salt.count,
                    hash,
                    UInt32(rounds),
                    derivedKeyBytes, keyByteCount)
            }
        }
        if (derivationStatus != 0) {
            print("Error: \(derivationStatus)")
            return nil;
        }

        return derivedKeyData
    }
}
